docs/src/dev/langref.md

```diff
@@ -1,4 +1,286 @@
+&<string>
+&<string>:<width>
+&<string>+<offset>
+&_
+&_:<width>
+```
+
+For example:
+
+```
+&"pin"
+&"gpio":8
+&"gpio"+3
+&_
+&_:16
+```
+
+I/O identifiers start with a `&` sigil followed by a non-empty [string](#string) (in which case the I/O identifier is *named*) or a `_` character (in which case it is *floating*). No whitespace is allowed between the sigil and the string or `_` character, or any of the following characters for the cases below.
+
+Named I/O identifiers may be followed by a `:` character and a non-empty sequence of `0`..`9` digits, which denotes the width of the declaration or the reference.
+
+Named I/O identifiers may be followed by a `+` character and a non-empty sequence of `0`..`9` digits, which denotes a 1-wide reference at the given offset.
+
+Floating I/O identifiers may be followed by a `:` character and a non-empty sequence of `0`..`9` digits, which denotes the width of the reference. They are used in [concatenations](#concat) to declare a partial connection to an I/O port.
+
+
+### Cell identifiers {#cell}
+
+```
+%<index>
+%<index>:<width>
+%<index>+<offset>
+%<index>+<offset>:<width>
+%<index>:_
+```
+
+For example:
+
+```
+%0
+%1:4
+%1+0
+%1+2:2
+%2:0
+%3:_
+```
+
+Cell identifiers start with a `%` sigil followed by a non-empty sequence of `0`..`9` digits, optionally followed by a `+` character and a non-empty sequence of `0`..`9` digits (denoting an offset into the wide output of a cell), optionally followed by a `:` character and a non-empty sequence of `0`..`9` digits (denoting the width of a cell declaration or reference). No whitespace is allowed between any of the characters.
+
+A *placeholder* cell identifier starts with a `%` sigil followed by a non-empty sequence of `0`..`9` digits, a `:` character, and a `_` character. No whitespace is allowed between any of the characters. The placeholder cell identifier is only used when declaring cells with multiple outputs.
+
+<div class="warning">
+
+While the in-memory IR has the property that a cell with a wide output occupies a contiguous range of cell indices (i.e. that `%0+1` and `%1` always refer to the same net), the same is not true in the text IR. Instead, each unique `<index>` can refer to a distinct cell with an output of any width.
+
+To avoid confusion, the IR printer ensures that in references, `<index>` is equal to the lowest index occupied by a cell with a wide output (i.e. it will never format `%0+1` as `%1`).
+
+</div>
+
+
+### Concatenations {#concat}
+
+```
+[ <part>... ]
+```
+
+For example:
+
+```
+[]
+[&"gpio":1 &"gpio":0]
+[&"pin" &_]
+[1001]
+[10 01]
+[%0 %1:4 000]
+[ XXXX ]
+```
+
+Concatenations are delimited by `[` and `]` characters on either end and contain zero or more parts delimited by [whitespace](#whitespace). The first part of the concatenation in the source code order corresponds to the most significant positions of the resulting reference. The two types of concatenations are *I/O concatenations* and *value concatenations*.
```

Comment:
Perhaps more succinctly:

> The syntactically first part of the concatenation corresponds to the most significant positions of the resulting reference.

Reply:
I wrote it that way first ("the lexically first") but I feel that spelling it out may be useful to readers unfamiliar with PL jargon.